### ISSUE TITLE
could not find the file of checkpoint_best_valid.pth and the config of detector.backbone_name of caption

### ISSUE THREAD

**Issue by verigle** [davidnvq/grit#4]
when I run the script of train_caption.py as follow command, the error occur：
```
python train_caption.py exp.name=caption_4ds model.detector.checkpoint=4ds_detector_path
```
> could not find the file : checkpoint_best_valid.pth

I found the code like this 

```
checkpoint = torch.load('checkpoint_best_valid.pth', map_location='cpu')
```

but where save the file of 'checkpoint_best_valid.pth' ?

**Comment by verigle** [comment]
I only have one GPU on one node, so  I set config file as default with rank = 0,
but I found the code like this :
```
        if rank == 1:
            best_cider_val = evaluate_metrics(
                model,
                optimizers,
                dataloader=dataloaders['valid_dict'],
                text_field=text_field,
                epoch=epoch,
                split='valid',
                config=config,
                train_res=train_res,
                writer=writer,
                best_cider=best_cider_val,
                which=phase,
                scheduler=scheduler,
            )

        if rank == 0:
            best_cider_test = evaluate_metrics(
                model,
                optimizers,
                dataloader=dataloaders['test_dict'],
                text_field=text_field,
                epoch=epoch,
                split='test',
                config=config,
                train_res=train_res,
                writer=writer,
                best_cider=best_cider_test,
                which=phase,
                scheduler=scheduler,
            )

```

the metrics of valid  will be  evaluated only rank == 1, is the reason of why could not find the file of checkpoint_best_valid.pth ?

**Comment by verigle** [comment]
counld not find config of  detector.backbone_name in coco_config.yaml,
I find a config like it in annother file in configs/detection/train_config.yaml:
```
model:
  has_attr_head: False

  backbone: 
    backbone_name: 'swin_base_win7_384_22k' 
```

can I use the config of detection/train_config.yaml to caption/coco_config.yaml ?

**Comment by verigle** [comment]
#6 #

**Comment by davidnvq** [comment]
Thank you for reporting the bug. I'm back to clean and refactor it again.

**Comment by davidnvq** [comment]
@verigle 
> but where save the file of 'checkpoint_best_valid.pth' ?
As we use hydra, the code will save the file in the working directory. And the working directory is located in the `dir` specified in the config file:
```
hydra:
  run:
    dir: ${oc.env:HOME}/checkpoints/grit/coco/${exp.name}
```

**Comment by davidnvq** [comment]
> I only have one GPU on one node, so I set config file as default with rank = 0, but I found the code like this :

You are true, I suppose that the training is performed on several GPUs (one or several nodes). I found that the evaluation code can't be evaluated in parallel. So, it is better to evaluate the test split on 1 GPU, and the validation split on another GPU. That's why I dedicate GPU 0 and GPU 1 for this purpose.
To fix it with one GPU, you can change the `if rank == 1:` by `if rank == 0`, then you will do evaluation on 2 splits sequentially. It will take time.

BTW, I am not sure that training with small batch size will affect the performance or not. If you do not have many GPUs, you can accumulate the gradient (e.g., [this tutorial](https://kozodoi.me/python/deep%20learning/pytorch/tutorial/2021/02/19/gradient-accumulation.html#:~:text=Simply%20speaking%2C%20gradient%20accumulation%20means,might%20find%20this%20tutorial%20useful.)). In our experiments, the overall `batch size = 16 x 8`.

**Comment by davidnvq** [comment]
> counld not find config of detector.backbone_name in coco_config.yaml,

Yes, we indeed use only ` 'swin_base_win7_384_22k'` as our backbone, that's why I decided to remove this from the config file. 

> can I use the config of detection/train_config.yaml to caption/coco_config.yaml ?

This config is actually used to train the detector, not for training on image caption task although they share similar hyperparameter settings.


> pull request of add config of detector.backbone_name in coco_config.yaml https://github.com/davidnvq/grit/pull/6

Let me resolve your PR. As the code is still messy, I will clean it based on your recommendation. Thanks.

**Comment by davidnvq** [comment]
It is seen that I was intentionally logging the backbone name in this line https://github.com/davidnvq/grit/blob/6ac14f85ace183fd22f894570ff7d6ac1bb6c764/engine/caption_engine.py#L119.
But  as I said, only `swin_base` is used. Then fixing this line is ok.

**Comment by davidnvq** [comment]
Feel free to open new issues. I close this since most issues seem to be resolved.
